src/compiler/scala/tools/nsc/typechecker/Namers.scala

```diff
@@ -1070,7 +1070,16 @@ trait Namers extends MethodSynthesis {
 
       val defnTpe = widenIfNecessary(tree.symbol, rhsTpe, pt)
       tree.tpt defineType defnTpe setPos tree.pos.focus
-      tree.tpt.tpe
+      val tpe = tree.tpt.tpe
+      // if enabled, validate that the now inferred val or def type isn't PwS
+      if (settings.warnInferProductWithSerializable && context.reportErrors) {
+        tpe match {
+          case RefinedType(ProductRootTpe :: SerializableTpe :: _, _) =>
+            reporter.warning(tree.pos, s"a type was inferred to be `$tpe`; this may indicate a programming error")
+          case _ =>
+        }
+      }
+      tpe
```

Comment:
The Namer check is for vals and defs, while the Infer check is for generic parameters. I can't find a suitable place in the compiler that would work for both situations without any duplication unfortunately, though that's mainly due to my unfamiliarity with the codebase.